### DISCUSSION TITLE
Synchronize the HTML-Preview

### DISCUSSION THREAD

**Discussion by Sci-Mon** [KDE/ghostwriter#858] [Ideas]
A synchronization of the HTML preview would be a fine thing. 
E.g. take the actual shown mardown-text of the editor and render it into HTML-Preview. You might be able to even delete the scrollbar because not needed anymore...

I've seen this a windows app (was that markdown pad?)

Another way might be to scroll the preview automatically to the correct position, however that can be solved.

I see that this might not be a simple task because the editor-font and the preview-font may differ...

**Comment by ewen-r** [comment]
+1 to this suggestion.
Unless I am missing something, the preview window and editor window positions are not in sync and I would very much like to have that.

**Comment by AlexShyXie** [comment]
+1 to this suggestion.

**Comment by actuallynotjohndoe** [comment]
+1

**Comment by Ardemus** [comment]
Agreed. It's worse for me on 24.02.2. In addition to what you describe, the preview scrolls itself all the way up when I make changes near the bottom of the markdown so it scrolls *away* from the part of the document I'm editing and I have to keep manually scrolling back down.

However the feature requires some non-trivial design decisions. The markdown and preview elements necessarily occupy different amounts of vertical space, and it's not consistent. In this screen shot both sides are synced at the top of the window (not captured):

![Screenshot 2024-05-26 001](https://github.com/KDE/ghostwriter/assets/16529978/7dbca858-a38a-4855-9b4f-91760b1f42d9)

So, which part of the markdown should align with the preview?

- The top of the visible markdown? 
- The markdown at the center of the screen?
- The markdown at the cursor?
- The most recent Header?
- The top of the current list/code block/paragraph?
- Some combination / heirarchy of conditions?
- Some sort of kerning-like operation to subtly change text size and line spacing so they do line up?
    - Do you allow adjustments to the preview or keep it WYSIWYG?
    - Do you go to extremes if necessary, or stop at a certain point?
- Break WYSIWYG in the preview by adding vertical space to both sides as needed so the top of each element lines up (like some diff utilities)?
- Keep the WYSIWYG but cut both sides into strips and insert gaps that are obviously not part of either?
- ...

Do you shrink one display area so they both contain the same content?
Do you use lines to connect the source to the preview like other diff utilities?

I definitely want them to stay synced, but it'd take work to decide *how* they sync.

**Reply by Sci-Mon** [reply]
I really appreciate your reply and the deep thoughts about the design decisions. I can imagine that it is almost not possible to realize a perfect WYSIWYG editor including the HTML preview.

I you have a look at this [screenshot](https://www.snapfiles.com/screenshots/markdownpad.htm), both editor and preview window have a scrollbar. It seems that the preview is scrolling only a little because the preview font is probably a bit too large. (But you might leave that issue to the users to choose their preferred font- and size)

Another difficult task would be to display the second half of a nested list of the visible markdown... Probably, it makes still sense to keep things simple and stupid at first and probably code such details later on.

For me it would be most natural to show the top of the visible markdown in synch and scroll the preview if needed. 
Another idea might be to just 'zoom' the preview somehow, until no scrolling is needed - although I cannot think of how to solve that exactly... Probably dynamically select the proper fontsize...

**Reply by distinctjuggle** [reply]
It seems pretty straightforward to me that the best solution is just matching the top line and not having concern for anything else.  Anything else would most importantly provide a very weird user experience, and to a lesser importance would have difficulties with edge cases as already described.  

I see no reason against matching the top of the visible markdown and not worrying about any cutoff.  If scrolling from the source worked like VSCode where you could scroll below the lowest line, this wouldn't present an issue unless I'm overlooking something else.  I can't imagine this is too terribly hard of a feature to implement, and if other approaches are desired in the future that could be refined later.

To me, this is pretty core functionality for a markdown editor with a preview.  It's incredibly harder to work without this feature.  It'd be great if this could see some attention in the meantime, as this is a pretty old feature request.

**Comment by AFlowOfCode** [comment]
This would be a lovely feature for a lovely editor.

I agree that some design thought would be required, but it could also be implemented in a simplistic way for uncomplicated situations as an opt-in feature. I don't feel like it has to strictly match up line to line to be more helpful than it is now. It could be labeled "beta" (especially if further refinement was being planned), or just "basic" to keep expectations low.
